tests/std/tests/P0466R5_layout_compatibility_and_pointer_interconvertibility_traits/test.cpp

```diff
@@ -0,0 +1,200 @@
+        struct S0 {
+            int v1;
+            int v2;
+        };
+
+        struct S1 {
+            S0 s1;
+            int v3;
+        };
+
+        struct S2 {
+            S0 s1;
+            int v2;
+        };
+
+        struct S3 {
+            S0 s1;
+            int v2;
+            int v3;
+        };
+
+        struct S4 {
+            int v1;
+
+        private:
+            int v2;
+        };
+
+        struct S5 {
+            int v1;
+
+        private:
+            int v2;
+        };
+
+        enum E1 { e1, e2, e3, e4 };
+        enum E2 : int { e5 };
+        enum E3 : unsigned int { e6, e7, e8 };
+        enum class E4 : unsigned int { no, yes };
+        enum class E5 { zero, fortytwo = 42 };
+
+        ASSERT(is_layout_compatible_v<int, int>);
+        ASSERT(is_layout_compatible_v<E1, E2>);
+        ASSERT(is_layout_compatible_v<E3, E4>);
+        ASSERT(is_layout_compatible_v<E5, E1>);
+        ASSERT(is_layout_compatible_v<const E1, E2>);
+        ASSERT(is_layout_compatible_v<volatile E3, const E4>);
+        ASSERT(is_layout_compatible_v<S1, volatile S2>);
+        ASSERT(is_layout_compatible_v<S1, S2>);
+        ASSERT(is_layout_compatible_v<S4, S4>);
+        ASSERT(is_layout_compatible_v<const volatile S4, S4>);
+
+        ASSERT(!is_layout_compatible_v<int, char>);
+        ASSERT(!is_layout_compatible_v<E1, E3>);
+        ASSERT(!is_layout_compatible_v<E2, E4>);
+        ASSERT(!is_layout_compatible_v<S1, S3>);
+        ASSERT(!is_layout_compatible_v<S4, S5>);
+    }
+
+    // is_pointer_interconvertible_base_of tests
+    {
+        class A {};
+        class B : public A {};
+        class C : public A {
+            int : 0;
+        };
+// Disable warning C4408: anonymous union did not declare any data members
+#pragma warning(push)
+#pragma warning(disable : 4408)
+        class D : public A {
+            union {};
+        };
+#pragma warning(pop)
+        class A1 : public A {};
+        class A2 : public A {};
+        class A3 : public A2 {};
+        class A4 : public A1, public A3 {};
+
+        ASSERT(is_pointer_interconvertible_base_of_v<A, A>);
+        ASSERT(is_pointer_interconvertible_base_of_v<A, B>);
+        ASSERT(is_pointer_interconvertible_base_of_v<A, const B>);
+        ASSERT(is_pointer_interconvertible_base_of_v<A, C>);
+        ASSERT(is_pointer_interconvertible_base_of_v<A, volatile C>);
+        ASSERT(is_pointer_interconvertible_base_of_v<volatile A, const C>);
+        ASSERT(is_pointer_interconvertible_base_of_v<A, D>);
+        ASSERT(is_pointer_interconvertible_base_of_v<A, A1>);
+        ASSERT(is_pointer_interconvertible_base_of_v<A, A2>);
+        ASSERT(is_pointer_interconvertible_base_of_v<A, A3>);
+        ASSERT(is_pointer_interconvertible_base_of_v<A2, A3>);
+
+        ASSERT(!is_pointer_interconvertible_base_of_v<A1, A2>);
+        ASSERT(!is_pointer_interconvertible_base_of_v<A4, A>);
+        ASSERT(!is_pointer_interconvertible_base_of_v<B, C>);
+    }

```

Comment:
The Condition for `is_pointer_interconvertible_base_of` is:

> `Derived` is unambiguously derived from `Base` without regard to cv-qualifiers, and each object of type `Derived` is pointer-interconvertible (6.8.3) with its `Base` subobject, or `Base` and `Derived` are not unions and name the same class type without regard to cv-qualifiers.

So the following cases aren't being tested:

* Non-`public` inheritance (as WG21-P0466 notes, "Here the derivation relationship need not be public, as pointer-interconvertibility does not respect access control.")
* `<int, int>` should be `false` (this tests "the same class type")
* `<MyUnion, MyUnion>` should be `false` (this tests "are not unions")

The requirements are:

> If `Base` and `Derived` are non-union class types and are not (possibly cv-qualified versions of) the same type, `Derived` shall be a complete type.

This indicates further test cases, where the compiler can confidently report `false` or `true` even when an input is incomplete:

* `<void, void>` should be `false`
* `<int, IncompleteClass>` should be `false`
* `<MyUnion, IncompleteClass>` should be `false`
* `<const IncompleteClass, IncompleteClass>` should be `true`